models/Image.js

```diff
@@ -1,76 +1,153 @@
+  const targets = {};
+
+  resizedOpts.forEach(resizedOpt => {
+    let [resolution, resizeWidth] = resizedOpt;
+    let resizeHeight = 0;
+
+    if (resizeWidth < originWidth) {
+      resizeHeight = Math.round((resizeWidth / originWidth) * originHeight);
+    } else {
+      resizeWidth = originWidth;
+      resizeHeight = originHeight;
+    }
+
+    targets[resolution] = {
+      url: `${resizeImageURLPrefix}${filenameWithoutExt}-${resolution}.${ext}`,
+      width: resizeWidth,
+      height: resizeHeight,
+    };
+  });
+
+  return targets;
+}
 
 Image.add({
-    image: {
-        type: Types.GcsImage,
-        initial: true,
-        autoCleanup: true,
-        datePrefix: 'YYYYMMDDHHmmss',
-        // TODO move these settings to config
-        bucket: bucket,
-        destination: 'images/',
-        publicRead: true,
-        resize: resizeImage,
-        resizeOpts: [{
-            target: 'desktop',
-            width: 2000,
-            height: null,
-            options: {}
-        }, {
-            target: 'tablet',
-            width: 1200,
-            height: null,
-            options: {}
-        }, {
-            target: 'mobile',
-            width: 800,
-            height: null,
-            options: {}
-        }, {
-            target: 'w400',
-            width: 400,
-            height: null,
-            options: {}
-        }, {
-            target: 'tiny',
-            width: 150,
-            height: null,
-            options: {}
-        }],
-        extractIPTC: extractIPTC
-    },
-    description: {
-        type: String,
-        index: true
-    },
-    copyright: {
-        type: Types.Select,
-        options: 'Creative-Commons, Copyrighted',
-        default: 'Copyrighted', index: true
-    },
-    keywords: {
-        type: String,
-        index: true
-    },
-    sale: {
-        type: Boolean,
-        index: true
+  image: {
+    type: Types.GcsImage,
+    initial: true,
+    autoCleanup: true,
+    datePrefix: 'YYYYMMDDHHmmss',
+    bucket: bucket,
+    destination: '/',
+    publicRead: false,
+    hooks: {
+      postUpload: async (imageData) => {
+        if (typeof imageData !== 'object' || imageData === null) {
+          return imageData;
+        }
+
+        const { filepath, filename, height, width } = imageData;
+
+        if (typeof filename !== 'string' || filename === '') {
+          return imageData;
+        }
+
+        try {
+          await publishResizeTasks(filename);
+        } catch (err) {
+          throw errors.helpers.wrap(err, 'Error', 'Received error while publishing resize events.')
+        }
+
+        try {
+          const iptc = await extractIPTC(filepath);
+          imageData.iptc = iptc;
+        } catch (err) {
+          throw errors.helpers.wrap(err, 'Error', 'Error to extract iptc from images.')
+        }
+
+        try {
+          const resizedTargets = getResizeMetadata(filename, width, height);
+          imageData.resizedTargets = resizedTargets;
```

Comment:
What is the case that `getResizeMetadata` will incur error or exception?

Reply:
There is no exception in `getResizeMetadata`.
This try catch is just for safety.

Reply:
It seems to me that this will be over protection and result in redundant code sections. I check the legacy implementation in @twreporter/keystone does not have the guard either. This makes me confused at the first sight. If we do not import any 3rd party package within these lines, I suggest we remove them safely.